src/main/mondrian/rolap/RolapSchema.java

```diff
@@ -746,138 +746,180 @@ private NamedSet createNamedSet(MondrianDef.NamedSet xmlNamedSet) {
-                        } catch (IllegalArgumentException e) {
-                            throw Util.newError(
-                                "Illegal rollupPolicy value '"
-                                + hierarchyGrant.rollupPolicy
-                                + "'");
-                        }
-                    } else {
-                        rollupPolicy = Role.RollupPolicy.FULL;
-                    }
-                    role.grant(
-                        hierarchy, hierarchyAccess, topLevel, bottomLevel,
-                        rollupPolicy);
-                    for (MondrianDef.MemberGrant memberGrant
-                        : hierarchyGrant.memberGrants)
-                    {
-                        if (hierarchyAccess != Access.CUSTOM) {
-                            throw Util.newError(
-                                "You may only specify <MemberGrant> if "
-                                + "<Hierarchy> has access='custom'");
-                        }
-                        Member member = schemaReader.withLocus()
-                            .getMemberByUniqueName(
-                                Util.parseIdentifier(memberGrant.member),
-                                !ignoreInvalidMembers);
-                        if (member == null) {
-                            // They asked to ignore members that don't exist
-                            // (e.g. [Store].[USA].[Foo]), so ignore this grant
-                            // too.
-                            assert ignoreInvalidMembers;
-                            continue;
-                        }
-                        if (member.getHierarchy() != hierarchy) {
-                            throw Util.newError(
-                                "Member '" + member
-                                + "' is not in hierarchy '" + hierarchy + "'");
-                        }
-                        role.grant(
-                            member,
-                            getAccess(memberGrant.access, memberAllowed));
-                    }
+                if (member.getHierarchy() != hierarchy) {
+                    throw Util.newError(
+                        "Member '" + member
+                        + "' is not in hierarchy '" + hierarchy + "'");
                 }
+                role.grant(
+                    member,
+                    getAccess(memberGrant.access, memberAllowed));
             }
         }
-        role.makeImmutable();
-        return role;
+
+        if (memberStored == 0 && hierarchyAccess == Access.CUSTOM) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(
+                    "Rolling back grants of Hierarchy '"
+                    + hierarchy.getUniqueName()
+                    + "' to ALL, because it contains no "
+                    + "valid restricted members");
+            }
+            role.grant(hierarchy, Access.ALL, null, null, rollupPolicy);
```

Comment:
"ignore" is a bit of a misnomer.  In most cases we treat non-existent members as the null member, we don't truly ignore.

Regardless, the scenario I gave above seems realistic and worrying.  If a role has been defined which limits access to a member or set of members, I wouldn't think we'd ever want to assume they *really* intended to grant access to everything.  It seems more likely that the dimensional data changed in some way, but we shouldn't assume that dimensional changes imply broader access rights.  Where security is involved it's best to be conservative.